Process command line load vernaculars before command line Goptions

LasseBlaauwbroek opened the pull request:
This pull request changes the order in which the command line options like `-load-vernac-object` and `-set` are processed. Currently, command-line load vernaculars are processed before command-line options. This is a problem when a load vernacular loads a new plugin that defines new options.

Example:
Load.v
```coq
Declare ML Module "my_options_plugin".
```
Now, if I use the following command, a warning is issued that the option `My Option` does not exist.
```bash
coqtop --load-vernac-object Load -set "My Option = 8"
```
This pull request fixes this. While I'm at it, I also added the `-r` alias of `--load-vernac-object` that was already documented in `coqtop -h` but was not working.

I'm not exactly sure what the impact of this change may be. The delayed setting of options may change some behavior. For example, an object loaded through `--load-vernac-object` may behave differently now that the options are set after it. As such, this is not a pull request that I expect to be merged, but rather have a discussion about. I would think that the ideal solution would be to set recognized options before processing the command-line load vernaculars and then add the unrecognized options as soon as they are available. I'm not sure how to achieve this, though.
<!-- Keep what applies -->
**Kind:** bug fix.


<!-- If there is a user-visible change in coqc/coqtop/coqchk/coq_makefile behavior and testing is not prohibitively expensive: -->
<!-- (Otherwise, remove this line.) -->
- [ ] Added / updated test-suite
<!-- If this is a feature pull request / breaks compatibility: -->
<!-- (Otherwise, remove these lines.) -->
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Zimmi48 commented:
> I'm not exactly sure what the impact of this change may be. The delayed setting of options may change some behavior. For example, an object loaded through `--load-vernac-object` may behave differently now that the options are set after it. As such, this is not a pull request that I expect to be merged, but rather have a discussion about.

As this `-set` command-line flag is quite recent, the impact might be limited, as also demonstrated by our CI. This however, would need to be documented in a changelog entry and in the reference manual. I'll summon @SkySkimmer who introduced the feature to get his opinion on this change.

> I would think that the ideal solution would be to set recognized options before processing the command-line load vernaculars and then add the unrecognized options as soon as they are available.

Personally, this kind of dynamic semantics frightens me quite a bit. I would very much prefer the simple inversion that you propose in your PR. If the two orders are really important to have, then a solution could be to have the order of command-line flags matter.

LasseBlaauwbroek commented:
> a solution could be to have the order of command-line flags matter.
That sounds like an excellent solution to me!

SkySkimmer commented:
> a solution could be to have the order of command-line flags matter.

I think that's what I would prefer, our code isn't really setup for it though.

LasseBlaauwbroek commented:
If the consensus is to do do the ordering of command-line flags then I can give it a try to implement this next week. But if someone more knowledgeable can do it quickly that is fine by me to of course :-)

ejgallego commented:
Indeed use cases do exist for any of the two possible ordering choices.

Given that @LasseBlaauwbroek has a use case for this I'm OK taking this PR in; a wider refactoring of the whole init setup is in the works, for other reasons both jscoq and serapi have to use their own init system so the idea is to unify all them.

I think that the real problem here is the `-load-vernac-object` option. Indeed that's a very low-level option that skips the whole document model of Coq; I am wondering if we should not just remove it in the first place. @LasseBlaauwbroek , why are you using that option?

ejgallego commented:
> If the consensus is to do do the ordering of command-line flags then I can give it a try to implement this next week. But if someone more knowledgeable can do it quickly that is fine by me to of course :-)

I'm not sure there is such a consensus, personally I'm not convinced as I don't see the use case for it, and it greatly increases the set of possible traces to consider in testing etc...

I think before implementing, opening an issue discussing some examples and intended design would help.

LasseBlaauwbroek commented:
> @LasseBlaauwbroek , why are you using that option?

I'm using it to inject a custom plugin into existing Coq developments distributed through Opam for instrumentation purposes. So for me this is a very useful option, and I would hate to see it go. Note that I'm fine with either one of `-load-vernac-object` and `-load-vernac-source`, but I assume that the first one is faster. The only other way I know to inject instrumentation code is to use the `coqrc` file but `coq_makefile` invokes `coqc` with `-q` so that doesn't work (and with good reason of course).

If it is decided that the interplay between `-load-vernac-*` and `-set` is too complicated, I would rather see `-set` be removed since `-load-vernac-*` is much more general that `-set`. The only reason `-set` is useful is because it lets me dynamically set options to my injected plugin without having to write them to a `.v` file. But I could live without it.

> I think before implementing, opening an issue discussing some examples and intended design would help.

I don't currently have any use-case for the command-line flags ordering proposal, but it does seem like a sensible design to me if both `-load-vernac-*` and `-set` are to be supported. However, for now I would also be happy is my current pull-request got merged.

LasseBlaauwbroek commented:
> it greatly increases the set of possible traces to consider in testing etc...

I should also note that the ordering of different `-load-vernac-object` options is already important so it seems to me that by incorporating `-set` into this order would not greatly increase the existing complexity if at all.

ejgallego commented:
I was mainly wondering about whether the `-require` family of options would be more appropriate; it mainly boils down to whether you want to instrument after the document is initialized or not.

But indeed setting the options later makes a lot of sense so let's go ahead with this PR.

LasseBlaauwbroek commented:
In practice I am indeed using `-rfrom`, but I was not aware of any semantic difference from `-load-vernac-object` outside of the usual difference between `Require` and `From .. Require`. Can you explain this in more detail to me.

ejgallego commented:
> Can you explain this in more detail to me?

IIANM one runs before the document itself is initialized; the other runs as part of the initialization. In practice, the differences are minimal, other than the modules are loaded at slightly different phases.

I need to look at the code, but for example the `-require` family will likely require modules after the prelude has been loaded.

Zimmi48 commented:
> So for me this is a very useful option, and I would hate to see it go.

Hey, don't worry. We never remove features with known users without proposing a clear replacement.

> If it is decided that the interplay between `-load-vernac-*` and `-set` is too complicated, I would rather see `-set` be removed

I don't think that the interplay is the problem.

> In practice I am indeed using `-rfrom`

So this PR does not change just the behavior of `-set` w.r.t `-load-vernac-object` IIUC. The whole set of changed commands will need to be documented.

To summarize, IIUC:

- @LasseBlaauwbroek: the options that are useful to you are `-ri` / `-rifrom`.
- These options behave in a more principled way than `-load-vernac-object` / `-load-vernac-source`.
- The options that @ejgallego would like to see gone are precisely the latter, not the former. What is blocking their immediate removal @ejgallego?
- We need to decide which order the various `-ri` and `-set` are executed in. I still strongly believe that using the order of the command-line arguments would be the more sensible option. However, because this is complicated to achieve, we can accept that this PR just reverse the order as a first step, making `-ri` be executed before `-set`.

Also:
- Note that `-require` is deprecated because its name is confusing since it does a `Require Import`.
- Do not add the `-r` shortcut for `-load-vernac-object` since this command is planned for removal and that `-ri` is preferable (instead, just remove this shortcut from the help menu). I think it was @herbelin's plan to reintroduce `-require` and `-r` to mean standard `Require` sometime after the current `-require` is removed.

ejgallego commented:
> To summarize, IIUC:

Yeah sorry if the discussion was complex; my summary of the PR is that we should take it as is, modulo the documentation and couple minor tweaks.

> * These options behave in a more principled way than `-load-vernac-object` / `-load-vernac-source`.

Just to remark that's still not clear to me; here we are in murky territory, I called them low-level and wondered if we should remove them; on the other hand they have more "power" than `-ri` et friends.

> * The options that @ejgallego would like to see gone are precisely the latter, not the former. What is blocking their immediate removal @ejgallego?

Actually they cost very little to support and it seems they have uses; we would gain little. Main drawback indeed is low-level semantics.

> * We need to decide which order the various `-ri` and `-set` are executed in. I still strongly believe that using the order of the command-line arguments would be the more sensible option. However, because this is complicated to achieve, we can accept that this PR just reverse the order as a first step, making `-ri` be executed before `-set`.

Actually not so complicated tho; we just need to have the Stm initializer take a list of initialization command which could be a set option or a require.

But indeed if we do that it seems to me that `-load-vernac-object` becomes a bit redundant; only drawback tho is if you want to run before the prelude, load-vernac-object makes that easy; `-ri` and friends require `-noinit` then actually `-ri` the prelude.

LasseBlaauwbroek commented:
@Zimmi48 Your summary seems accurate. I updated the PR to remove the `-r` option. I also added a changelog entry (hopefully I did this right). Let me know where else this change needs to be documented.

> But indeed if we do that it seems to me that -load-vernac-object becomes a bit redundant; only drawback tho is if you want to run before the prelude, load-vernac-object makes that easy; -ri and friends require -noinit then actually -ri the prelude.

Couldn't we make it so that `-coqlib` is also relevant in the ordering? Then you can explicitly order everything.

ejgallego commented:
> Then you can explicitly order everything.

What role would `-coqlib` play here? It only determines include paths, not when modules are loaded.

Zimmi48 commented:
@LasseBlaauwbroek You didn't update the right changelog. This is a user-facing change. Please see the documentation at: https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog

You also need to update the documentation of the command-line arguments (located under `doc/sphinx/practical-tools/coq-commands.rst`).

LasseBlaauwbroek commented:
Well, you could kind of see this as a variant of `-rfrom` except specifically for the prelude I guess (although I realize that the connection is a bit vague). But my intention was that in the example
`coqtop -rfrom A X -coqlib dir -rfrom B Y`
the ordering is to first load `A.X`, then `Init.Prelude` and finally `B.Y`. If this is too vague, it could also be another option like `-rprelude`. But maybe just doing an `-ri Coq Prelude` is just as easy. Not sure.

Zimmi48 commented:
`coqtop -noinit -ri A X -ri Coq Prelude -ri B Y` would seem to me like the natural way to do this. Indeed, I agree with @ejgallego that `-coqlib` is kind of orthogonal to this.

LasseBlaauwbroek commented:
@Zimmi48 Should be done.

LasseBlaauwbroek commented:
Done. Now that I'm thinking about it, I agree that just removing `-load-vernac-object` is completely fine and the existing ordering options are fine. It's not like people are doing this kind of stuff every day. Then only the interleaving of `-ri` and `-set` remains. Should I try to implement this, or create an issue for discussion?

Zimmi48 commented:
> Should I try to implement this

@ejgallego seemed to think that this was doable, so if you're up to the task, that would be cool. Make sure you follow his guidance though:

>Actually not so complicated tho; we just need to have the Stm initializer take a list of initialization command which could be a set option or a require.

ejgallego commented:
> Actually not so complicated tho; we just need to have the Stm initializer take a list of initialization command which could be a set option or a require.

Indeed, as noted by @LasseBlaauwbroek we already are order-dependent on a few lists; just unify the two lists into one and that should work, using the typical OCaml sum type should do the job.

ejgallego commented:
Ups, I tried to merge but there is a `needs: squashing` tag, excuse me @LasseBlaauwbroek  , would you indeed mind squashing the commits to avoid history noise? I'll merge ASAP.

LasseBlaauwbroek commented:
Done. I'll look at the interleaving issue probably next week.
